etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/ruleengine/transformation/custom/MapLocalObservationCodes.java

```diff
@@ -0,0 +1,152 @@
+ * added to the observation as an additional observation coding with cwe-coding extension 'coding'.
+ * When converted to/from HL7, these codings correspond to OBX-3.4/5/6 for the local code, and
+ * OBX-3.1/2/3 for the LOINC/PLT code.
+ */
+public class MapLocalObservationCodes implements CustomFhirTransformation {
+    protected final Logger LOGGER = ApplicationContext.getImplementation(Logger.class);
+
+    private HashMap<String, IdentifierCode> codingMap;
+
+    public MapLocalObservationCodes() {
+        initMap();
+    }
+
+    @Override
+    public void transform(FhirResource<?> resource, Map<String, String> args) {
+        var bundle = (Bundle) resource.getUnderlyingResource();
+        var observations = HapiHelper.resourcesInBundle(bundle, Observation.class);
+
+        for (Observation obv : observations.toList()) {
+            var codingList = obv.getCode().getCoding();
+
+            if (codingList.size() == 1) {
+                var coding = codingList.get(0);
+
+                var cwe =
+                        coding.getExtensionByUrl(HapiHelper.EXTENSION_CWE_CODING)
+                                .getValue()
+                                .toString();
+
+                // Alt coding is HL7 OBX-3.4,5,6
+                if (Objects.equals(cwe, "alt-coding")
+                        && coding.getSystem().equals(HapiHelper.LOCAL_CODE_URL)) {
+
+                    // Look up the code in the hash map
+                    var identifier = codingMap.get(coding.getCode());
+
+                    if (identifier == null) {
+                        // The local code was not found in the mapping
+                        var msh41Identifier = HapiHelper.getMSH4_1Identifier(bundle);
+                        var msh41Value =
+                                msh41Identifier != null ? msh41Identifier.getValue() : null;
+
+                        LOGGER.logWarning(
+                                "Unmapped local code detected: '{}', from sender: '{}', message Id: '{}'",
+                                coding.getCode(),
+                                msh41Value,
+                                HapiHelper.getMessageControlId(bundle));
+                        continue;
+                    }
+
+                    // Create a new coding source and add it to the coding list
+                    var mappedCoding =
+                            new Coding(
+                                    urlForCodeType(identifier.codingSystem()),
+                                    identifier.code(),
+                                    identifier.display());
+                    mappedCoding.addExtension(
+                            HapiHelper.EXTENSION_CWE_CODING, new StringType("coding"));
+
+                    mappedCoding.addExtension(
+                            HapiHelper.EXTENSION_CODING_SYSTEM,
+                            new StringType(identifier.codingSystem()));
+
+                    codingList.add(0, mappedCoding);
```

Comment:
This will always have a coding list with 1 entry, is that the intent?

Reply:
Actually, the first entry will be moved to the next index [1], and the new entry will be in index [0], and this is the desired/expected behavior.